// @salt-ds/lab/dist-cjs/calendar/CalendarWeekHeader.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var react = require('react');
var utils = require('./internal/utils.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var CalendarWeekHeader$1 = require('./CalendarWeekHeader.css.js');
var formatDate = require('./formatDate.js');

const withBaseName = core.makePrefixer("saltCalendarWeekHeader");
const CalendarWeekHeader = react.forwardRef(function CalendarWeekHeader2({ className, locale = formatDate.getCurrentLocale(), ...rest }, ref) {
  const weekdaysShort = utils.daysForLocale("narrow", locale);
  const weekdaysLong = utils.daysForLocale("long", locale);
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-calendar-week-header",
    css: CalendarWeekHeader$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    "data-testid": "CalendarWeekHeader",
    className: clsx.clsx(withBaseName(), className),
    ref,
    ...rest,
    children: weekdaysShort.map((day, index) => /* @__PURE__ */ jsxRuntime.jsx("small", {
      "aria-hidden": true,
      className: withBaseName("dayOfWeek"),
      children: day
    }, weekdaysLong[index]))
  });
});

exports.CalendarWeekHeader = CalendarWeekHeader;
