// @salt-ds/lab/dist-cjs/combo-box-deprecated/internal/useComboBox.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var react = require('react');
require('../../list-deprecated/List.js');
require('../../list-deprecated/ListBase.js');
require('../../list-deprecated/ListItemBase.js');
require('../../list-deprecated/ListItem.js');
require('../../list-deprecated/ListItemContext.js');
require('../../list-deprecated/ListStateContext.js');
var useList = require('../../list-deprecated/useList.js');
var defaultItemToString = require('../../tokenized-input/internal/defaultItemToString.js');
var filterHelpers = require('../filterHelpers.js');
var usePopperStatus = require('./usePopperStatus.js');

const useComboBox = (props) => {
  const {
    initialOpen,
    initialSelectedItem,
    allowFreeText,
    displayedItemCount,
    virtualized,
    disabled,
    onBlur,
    onFocus,
    onChange,
    onSelect,
    onInputChange,
    onInputFocus,
    onInputBlur,
    onInputSelect,
    id: idProp,
    source: sourceProp = [],
    selectedItem: selectedItemProp,
    inputValue: inputValueProp,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    getFilterRegex = filterHelpers.getDefaultFilterRegex,
    itemToString = defaultItemToString.defaultItemToString,
    stringToItem: stringToItemProp = (value) => value ? value.trim() : value,
    InputProps: inputProps = {
      onBlur: onBlur || onInputBlur,
      onFocus: onFocus || onInputFocus,
      onChange: onInputChange,
      onSelect: onInputSelect
    },
    ListProps = {},
    ...restProps
  } = props;
  const id = core.useId(idProp);
  const inputId = `${id}-input`;
  const listId = `${id}-list`;
  const [inputValue, setInputValue] = core.useControlled({
    controlled: inputValueProp,
    default: "",
    name: "ComboBox",
    state: "inputValue"
  });
  const [selectionChanged, setSelectionChanged] = react.useState(false);
  const inputRef = react.useRef(null);
  const [allowAnnouncement, setAllowAnnouncement] = react.useState(false);
  const labels = react.useMemo(
    () => sourceProp.map(itemToString),
    [sourceProp, itemToString]
  );
  const source = react.useMemo(() => {
    if (inputValue == null ? void 0 : inputValue.trim().length) {
      const itemFilter = filterHelpers.getDefaultFilter(inputValue, getFilterRegex);
      return sourceProp.filter((item) => itemFilter(itemToString(item)));
    }
    return sourceProp;
  }, [inputValue, sourceProp, getFilterRegex, itemToString]);
  const itemTextHighlightPattern = react.useMemo(
    () => (inputValue == null ? void 0 : inputValue.trim().length) ? getFilterRegex(inputValue) : void 0,
    [inputValue, getFilterRegex]
  );
  const { focusedRef, state, helpers, listProps } = useList.useList({
    ...ListProps,
    source,
    disabled,
    virtualized,
    itemToString,
    initialSelectedItem,
    selectedItem: selectedItemProp,
    displayedItemCount,
    onChange,
    onSelect,
    id: listId,
    tabToSelect: true,
    disableFocus: true,
    disableMouseDown: true,
    "aria-labelledby": ariaLabelledBy
  });
  const {
    isFocusVisibleRef,
    onFocus: handleFocusVisible,
    onBlur: handleBlurVisible,
    ref: focusVisibleRef
  } = core.useIsFocusVisible();
  const handleFocusVisibleRef = core.useForkRef(focusVisibleRef, focusedRef);
  const handleInputRef = core.useForkRef(inputRef, handleFocusVisibleRef);
  const { "aria-activedescendant": ariaActiveDescendant, ...restListProps } = listProps;
  const { selectedItem, highlightedIndex } = state;
  const [quickSelection, setQuickSelection] = react.useState(false);
  const { setFocusVisible, setSelectedItem, setHighlightedIndex } = helpers;
  const reconcileInput = react.useCallback(() => {
    setInputValue(selectedItem == null ? "" : itemToString(selectedItem));
  }, [selectedItem, itemToString]);
  const selectInputValue = (event) => {
    const nextIndex = inputValue ? labels.indexOf(inputValue.trim()) : -1;
    const nextItem = nextIndex >= 0 ? sourceProp[nextIndex] : stringToItemProp(inputValue);
    if (onSelect) {
      onSelect(event, nextItem);
    }
    if (nextItem !== selectedItem) {
      setSelectedItem(nextItem);
      onChange == null ? void 0 : onChange(event, nextItem);
    }
  };
  const { isOpen: isListOpen, notifyPopper } = usePopperStatus.usePopperStatus({
    initialOpen
  });
  react.useEffect(() => {
    reconcileInput();
  }, [reconcileInput]);
  react.useEffect(() => {
    if (!isListOpen) {
      setHighlightedIndex(void 0);
      setQuickSelection(false);
    }
  }, [isListOpen, setHighlightedIndex]);
  const initHighlightedIndex = () => {
    setHighlightedIndex(selectedItem ? source.indexOf(selectedItem) : -1);
  };
  const handleInputFocus = (event) => {
    handleFocusVisible(event);
    if (isFocusVisibleRef.current) {
      setFocusVisible(true);
    }
    if (highlightedIndex == null) {
      initHighlightedIndex();
    }
    if (inputProps.onFocus) {
      inputProps.onFocus(event);
    }
    notifyPopper(event);
  };
  const handleInputBlur = (event) => {
    console.log("BLUR");
    handleBlurVisible();
    setAllowAnnouncement(false);
    if (allowFreeText) {
      selectInputValue(event);
    } else {
      reconcileInput();
    }
    if (restListProps.onBlur) {
      restListProps.onBlur(event);
    }
    if (inputProps.onBlur) {
      inputProps.onBlur(event);
    }
    notifyPopper(event);
  };
  const handleInputChange = (event) => {
    const newValue = event.target.value;
    setInputValue(newValue);
    setHighlightedIndex(void 0);
    setQuickSelection(newValue.length > 0 && !allowFreeText);
    if (newValue.length === 0) {
      setSelectedItem(void 0);
      onChange == null ? void 0 : onChange(event, null);
    }
    if (inputProps.onChange) {
      inputProps.onChange(event, "");
    }
  };
  const handleInputSelect = (event) => {
    console.log("handleInputSelect setSelectionChanged = true");
    setSelectionChanged(true);
    if (inputProps.onSelect) {
      inputProps.onSelect(event);
    }
  };
  const handleInputKeyDown = (event) => {
    if ("Escape" === event.key) {
      setHighlightedIndex(void 0);
      if (allowFreeText) {
        setInputValue("");
      } else {
        reconcileInput();
      }
    }
    if (inputProps.onKeyDown) {
      inputProps.onKeyDown(event);
    }
  };
  const handleFirstItemSelection = (event) => {
    if (!allowFreeText && event.key === "Enter" && quickSelection) {
      setSelectedItem(source[0]);
      onSelect == null ? void 0 : onSelect(event, source[0]);
      onChange == null ? void 0 : onChange(event, source[0]);
    }
  };
  const handleKeyDown = (event) => {
    if (usePopperStatus.isToggleList(event)) {
      notifyPopper(event);
      return;
    }
    if (["ArrowUp", "ArrowDown"].indexOf(event.key) !== -1 && highlightedIndex == null) {
      initHighlightedIndex();
    }
    handleFirstItemSelection(event);
    if ([" ", "Home", "End"].indexOf(event.key) === -1) {
      if (restListProps.onKeyDown) {
        restListProps.onKeyDown(event);
      }
      console.log(`handleKeyDown ${event.key} setSelectionCHanged = false`);
      setSelectionChanged(false);
    }
    setAllowAnnouncement("Backspace" !== event.key);
    handleInputKeyDown(event);
    notifyPopper(event);
  };
  const handleListClick = (event) => {
    console.log("handle list click");
    const inputEl = inputRef.current;
    if (inputEl != null) {
      inputEl.focus();
    }
    if (restListProps.onClick) {
      restListProps.onClick(event);
    }
    notifyPopper(event);
  };
  const mergedInputProps = {
    ...inputProps.inputProps,
    role: "combobox",
    "aria-owns": listId,
    "aria-label": ariaLabel,
    "aria-expanded": isListOpen,
    "aria-activedescendant": isListOpen && selectionChanged ? null : ariaActiveDescendant
  };
  return {
    inputRef: handleInputRef,
    listContext: {
      state,
      helpers
    },
    inputProps: {
      ...restProps,
      ...inputProps,
      disabled,
      allowAnnouncement,
      id: inputId,
      value: inputValue,
      onFocus: handleInputFocus,
      onBlur: handleInputBlur,
      onChange: handleInputChange,
      onKeyDown: handleKeyDown,
      onSelect: handleInputSelect,
      inputProps: mergedInputProps
    },
    listProps: {
      ...restListProps,
      source,
      itemToString,
      itemTextHighlightPattern,
      onClick: handleListClick,
      isListOpen: isListOpen && Boolean(source.length)
    }
  };
};

exports.useComboBox = useComboBox;
