// @salt-ds/lab/dist-cjs/slider/internal/SliderThumb.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var react = require('react');
var SliderContext = require('./SliderContext.js');
var useKeyDownThumb = require('./useKeyDownThumb.js');
var utils = require('./utils.js');

const withBaseName = core.makePrefixer("saltSliderThumb");
function SliderThumb(props) {
  const { index, activeThumb, setActiveThumb } = props;
  const [focussed, setFocussed] = react.useState(false);
  const { min, max, step, value, onChange, ariaLabel } = SliderContext.useSliderContext();
  const onKeyDown = useKeyDownThumb.useKeyDownThumb(min, max, step, value, onChange, index);
  const thumbValue = value[index];
  const thumbPosition = utils.getPercentage(min, max, thumbValue);
  const handlePointerOver = () => {
    if (activeThumb === void 0 && index !== null)
      setActiveThumb(index);
  };
  const handleFocus = () => {
    setFocussed(true);
    if (index !== null)
      setActiveThumb(index);
  };
  const handleBlur = () => {
    setFocussed(false);
    setActiveThumb(void 0);
  };
  const showTooltip = focussed || activeThumb === index;
  return /* @__PURE__ */ jsxRuntime.jsxs("div", {
    style: { left: `${thumbPosition}%` },
    className: withBaseName("container"),
    onPointerOver: handlePointerOver,
    onFocus: handleFocus,
    onBlur: handleBlur,
    children: [
      /* @__PURE__ */ jsxRuntime.jsxs("div", {
        className: clsx.clsx(withBaseName("tooltip"), {
          [withBaseName("tooltip-visible")]: showTooltip
        }),
        "aria-expanded": showTooltip,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx("svg", {
            className: withBaseName("tooltip-arrow"),
            "aria-hidden": "true",
            viewBox: "0 1 14 14",
            children: /* @__PURE__ */ jsxRuntime.jsx("path", {
              d: "M0,0 H14 L7,7 Q7,7 7,7 Z"
            })
          }),
          /* @__PURE__ */ jsxRuntime.jsx(core.Label, {
            children: value[index]
          })
        ]
      }),
      /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: withBaseName(),
        onKeyDown,
        role: "slider",
        "aria-valuemin": min,
        "aria-valuemax": max,
        "aria-valuenow": index === 1 ? value[1] : value[0],
        "aria-label": ariaLabel,
        "aria-orientation": "horizontal",
        tabIndex: 0
      })
    ]
  });
}

exports.SliderThumb = SliderThumb;
