// @salt-ds/lab/dist-cjs/query-input/internal/CategoryListItem.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var react = require('react');
require('../../list/Highlighter.js');
var ListItem = require('../../list/ListItem.js');
require('../../list/List.js');
require('../../common-hooks/collectionProvider.js');
require('../../common-hooks/keyUtils.js');
require('../../responsive/useResizeObserver.js');
require('../../list/VirtualizedList.js');
var CategoryListContext = require('./CategoryListContext.js');

const withBaseName = core.makePrefixer("saltCategoryListItem");
const CategoryListItem = function CategoryListItem2({ item: category, ...props }) {
  const textRef = react.useRef(null);
  const context = CategoryListContext.useCategoryListContext();
  const { ExpandGroupIcon } = core.useIcon();
  const textStyle = react.useMemo(
    () => ({
      minWidth: context.width
    }),
    [context.width]
  );
  return /* @__PURE__ */ jsxRuntime.jsxs(ListItem.ListItem, {
    ...props,
    label: category == null ? void 0 : category.name,
    children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", {
        ref: textRef,
        className: withBaseName("text"),
        style: textStyle,
        children: category == null ? void 0 : category.name
      }),
      /* @__PURE__ */ jsxRuntime.jsxs("div", {
        className: withBaseName("valuesContainer"),
        children: [
          /* @__PURE__ */ jsxRuntime.jsx("span", {
            children: "("
          }),
          /* @__PURE__ */ jsxRuntime.jsx("span", {
            className: withBaseName("values"),
            children: category == null ? void 0 : category.values.join(", ")
          }),
          /* @__PURE__ */ jsxRuntime.jsx("span", {
            children: ")"
          })
        ]
      }),
      /* @__PURE__ */ jsxRuntime.jsx(ExpandGroupIcon, {
        className: withBaseName("chevron")
      })
    ]
  });
};

exports.CategoryListItem = CategoryListItem;
