// @salt-ds/lab/dist-es/date-picker/DatePickerRangeInput.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer, useIcon, useForkRef, useControlled, Button } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useRef, useCallback, useEffect } from 'react';
import '../date-input/DateInputSingle.js';
import { DateInputRange } from '../date-input/DateInputRange.js';
import '@internationalized/date';
import { useDatePickerContext } from './DatePickerContext.js';
import { useDatePickerOverlay } from './DatePickerOverlayProvider.js';

const withBaseName = makePrefixer("saltDatePickerRangeInput");
const DatePickerRangeInput = forwardRef(function DatePickerRangeInput2(props, ref) {
  const {
    className,
    endInputProps: endInputPropsProp,
    startInputProps: startInputPropsProp,
    onKeyDown,
    defaultValue,
    value: valueProp,
    onChange,
    onDateValueChange,
    ...rest
  } = props;
  const { CalendarIcon } = useIcon();
  const {
    state: { selectedDate, disabled, readOnly, cancelled, locale, timeZone },
    helpers: { setSelectedDate }
  } = useDatePickerContext({ selectionVariant: "range" });
  const {
    state: { open, floatingUIResult },
    helpers: { getReferenceProps, setOpen }
  } = useDatePickerOverlay();
  const inputRef = useForkRef(ref, floatingUIResult == null ? void 0 : floatingUIResult.reference);
  const prevState = useRef();
  const [value, setValue] = useControlled({
    controlled: valueProp,
    default: defaultValue,
    name: "DatePickerRangeInput",
    state: "dateValue"
  });
  const handleCalendarButton = useCallback(() => {
    setOpen(!open);
  }, [open, setOpen]);
  const handleDateChange = useCallback(
    (_event, newDate, error) => {
      setSelectedDate(newDate, error);
    },
    [setSelectedDate]
  );
  const handleDateValueChange = (newDateValue, isFormatted) => {
    setValue(newDateValue);
    onDateValueChange == null ? void 0 : onDateValueChange(newDateValue, isFormatted);
  };
  useEffect(() => {
    if (open) {
      prevState.current = { date: selectedDate, value };
    }
  }, [open]);
  useEffect(() => {
    var _a, _b;
    if (cancelled) {
      setValue((_a = prevState == null ? void 0 : prevState.current) == null ? void 0 : _a.value);
      setSelectedDate(((_b = prevState == null ? void 0 : prevState.current) == null ? void 0 : _b.date) || null, {
        startDate: false,
        endDate: false
      });
    }
  }, [cancelled]);
  const startInputProps = {
    onKeyDown: (event) => {
      var _a;
      if (event.key === "ArrowDown") {
        setOpen(true);
      }
      (_a = startInputPropsProp == null ? void 0 : startInputPropsProp.onKeyDown) == null ? void 0 : _a.call(startInputPropsProp, event);
    },
    ...startInputPropsProp
  };
  const endInputProps = {
    onKeyDown: (event) => {
      var _a;
      if (event.key === "ArrowDown") {
        setOpen(true);
      }
      (_a = endInputPropsProp == null ? void 0 : endInputPropsProp.onKeyDown) == null ? void 0 : _a.call(endInputPropsProp, event);
    },
    ...endInputPropsProp
  };
  return /* @__PURE__ */ jsx(DateInputRange, {
    value: value || { startDate: "", endDate: "" },
    locale,
    timeZone,
    className: clsx(withBaseName(), className),
    ref: inputRef,
    date: selectedDate,
    startInputProps: getReferenceProps(startInputProps),
    endInputProps: getReferenceProps(endInputProps),
    readOnly,
    onDateChange: handleDateChange,
    onDateValueChange: handleDateValueChange,
    onChange,
    endAdornment: /* @__PURE__ */ jsx(Button, {
      appearance: "transparent",
      sentiment: "neutral",
      onClick: handleCalendarButton,
      disabled,
      "aria-label": "Open Calendar",
      children: /* @__PURE__ */ jsx(CalendarIcon, {})
    }),
    ...rest
  });
});

export { DatePickerRangeInput };
