// @salt-ds/lab/dist-es/date-picker/DatePickerRangePanel.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { getLocalTimeZone, startOfMonth, today, isSameMonth, endOfMonth } from '@internationalized/date';
import { makePrefixer, useControlled, StackLayout, FlexItem, FormFieldHelperText, FlexLayout, FormFieldContext } from '@salt-ds/core';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { forwardRef, useState, useCallback } from 'react';
import { Calendar } from '../calendar/Calendar.js';
import { CalendarNavigation } from '../calendar/CalendarNavigation.js';
import { CalendarWeekHeader } from '../calendar/CalendarWeekHeader.js';
import { CalendarDateGrid } from '../calendar/CalendarDateGrid.js';
import '../calendar/internal/CalendarContext.js';
import { getCurrentLocale } from '../calendar/formatDate.js';
import '../calendar/useCalendarSelection.js';
import { useDatePickerContext } from './DatePickerContext.js';
import css_248z from './DatePickerPanel.css.js';

function getFallbackVisibleMonths(selectedDate, timeZone) {
  const createConsecutiveRange = (date) => [
    startOfMonth(date),
    startOfMonth(date).add({ months: 1 })
  ];
  if (selectedDate == null ? void 0 : selectedDate.startDate) {
    const { startDate, endDate } = selectedDate;
    if (endDate) {
      return isSameMonth(startDate, endDate) ? createConsecutiveRange(startDate) : [startOfMonth(startDate), startOfMonth(endDate)];
    }
    return createConsecutiveRange(startDate);
  }
  const currentMonth = startOfMonth(today(timeZone));
  return [currentMonth, currentMonth.add({ months: 1 })];
}
const withBaseName = makePrefixer("saltDatePickerPanel");
const DatePickerRangePanel = forwardRef(function DatePickerRangePanel2(props, ref) {
  const {
    className,
    defaultStartVisibleMonth: defaultStartVisibleMonthProp,
    startVisibleMonth: startVisibleMonthProp,
    onStartVisibleMonthChange,
    defaultEndVisibleMonth: defaultEndVisibleMonthProp,
    endVisibleMonth: endVisibleMonthProp,
    onEndVisibleMonthChange,
    helperText,
    onSelect,
    StartCalendarProps: StartCalendarPropsProp,
    StartCalendarNavigationProps,
    StartCalendarWeekHeaderProps,
    StartCalendarDataGridProps,
    EndCalendarProps: EndCalendarPropsProp,
    EndCalendarNavigationProps,
    EndCalendarWeekHeaderProps,
    EndCalendarDataGridProps,
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-date-picker-range-panel",
    css: css_248z,
    window: targetWindow
  });
  const {
    state: {
      selectedDate,
      timeZone = getLocalTimeZone(),
      minDate = startOfMonth(today(timeZone)),
      maxDate = minDate.add({ months: 1 }),
      locale = getCurrentLocale()
    },
    helpers: { setSelectedDate }
  } = useDatePickerContext({ selectionVariant: "range" });
  const [hoveredDate, setHoveredDate] = useState(null);
  const [[fallbackStartVisibleMonth, fallbackEndVisibleMonth]] = useState(
    () => getFallbackVisibleMonths(selectedDate, timeZone)
  );
  const [startVisibleMonth, setStartVisibleMonth] = useControlled({
    controlled: startVisibleMonthProp,
    default: defaultStartVisibleMonthProp || fallbackStartVisibleMonth,
    name: "DatePickerRangePanel",
    state: "startVisibleMonth"
  });
  const [endVisibleMonth, setEndVisibleMonth] = useControlled({
    controlled: endVisibleMonthProp,
    default: defaultEndVisibleMonthProp || fallbackEndVisibleMonth,
    name: "DatePickerRangePanel",
    state: "endVisibleMonth"
  });
  const handleSelectedDateChange = useCallback(
    (event, newDate) => {
      setSelectedDate(newDate, { startDate: false, endDate: false });
      onSelect == null ? void 0 : onSelect(event, newDate);
    },
    [onSelect, setSelectedDate]
  );
  const handleHoveredStartDateChange = useCallback(
    (event, newHoveredDate) => {
      var _a;
      setHoveredDate(newHoveredDate);
      if (newHoveredDate && (StartCalendarPropsProp == null ? void 0 : StartCalendarPropsProp.onHoveredDateChange)) {
        (_a = StartCalendarPropsProp.onHoveredDateChange) == null ? void 0 : _a.call(StartCalendarPropsProp, event, newHoveredDate);
      }
    },
    [StartCalendarPropsProp == null ? void 0 : StartCalendarPropsProp.onHoveredDateChange]
  );
  const handleHoveredEndDateChange = useCallback(
    (event, newHoveredDate) => {
      setHoveredDate(newHoveredDate);
      if (newHoveredDate && (EndCalendarPropsProp == null ? void 0 : EndCalendarPropsProp.onHoveredDateChange)) {
        EndCalendarPropsProp.onHoveredDateChange(event, newHoveredDate);
      }
    },
    [EndCalendarPropsProp == null ? void 0 : EndCalendarPropsProp.onHoveredDateChange]
  );
  const handleStartVisibleMonthChange = useCallback(
    (event, newVisibleMonth) => {
      setStartVisibleMonth(newVisibleMonth);
      if (newVisibleMonth.compare(endVisibleMonth) >= 0) {
        setEndVisibleMonth(newVisibleMonth.add({ months: 1 }));
      }
      onStartVisibleMonthChange == null ? void 0 : onStartVisibleMonthChange(event, newVisibleMonth);
    },
    [endVisibleMonth, onStartVisibleMonthChange]
  );
  const handleEndVisibleMonthChange = useCallback(
    (event, newVisibleMonth) => {
      setEndVisibleMonth(newVisibleMonth);
      if (newVisibleMonth.compare(startVisibleMonth) <= 0) {
        setStartVisibleMonth(
          startOfMonth(newVisibleMonth.subtract({ months: 1 }))
        );
      }
      onEndVisibleMonthChange == null ? void 0 : onEndVisibleMonthChange(event, newVisibleMonth);
    },
    [startVisibleMonth, onEndVisibleMonthChange]
  );
  function getHoveredDate(date, hoveredDate2) {
    return date && hoveredDate2 && (hoveredDate2 == null ? void 0 : hoveredDate2.compare(endOfMonth(date))) > 0 ? endOfMonth(date) : hoveredDate2;
  }
  const StartCalendarProps = {
    visibleMonth: startVisibleMonth,
    hoveredDate: getHoveredDate(selectedDate == null ? void 0 : selectedDate.startDate, hoveredDate),
    selectedDate,
    onHoveredDateChange: handleHoveredStartDateChange,
    onVisibleMonthChange: handleStartVisibleMonthChange,
    onSelectedDateChange: handleSelectedDateChange,
    hideOutOfRangeDates: true,
    minDate,
    maxDate,
    locale,
    timeZone,
    ...StartCalendarPropsProp
  };
  const EndCalendarProps = {
    visibleMonth: endVisibleMonth,
    hoveredDate,
    selectedDate,
    onHoveredDateChange: handleHoveredEndDateChange,
    onVisibleMonthChange: handleEndVisibleMonthChange,
    onSelectedDateChange: handleSelectedDateChange,
    hideOutOfRangeDates: true,
    minDate,
    maxDate,
    locale,
    timeZone,
    ...EndCalendarPropsProp
  };
  return /* @__PURE__ */ jsxs(StackLayout, {
    separators: true,
    gap: 0,
    className: clsx(className, withBaseName("container")),
    ref,
    ...rest,
    children: [
      helperText && /* @__PURE__ */ jsx(FlexItem, {
        className: withBaseName("header"),
        children: /* @__PURE__ */ jsx(FormFieldHelperText, {
          children: helperText
        })
      }),
      /* @__PURE__ */ jsx(FlexLayout, {
        gap: 0,
        children: /* @__PURE__ */ jsxs(FormFieldContext.Provider, {
          value: {},
          children: [
            /* @__PURE__ */ jsxs(Calendar, {
              selectionVariant: "range",
              ...StartCalendarProps,
              children: [
                /* @__PURE__ */ jsx(CalendarNavigation, {
                  ...StartCalendarNavigationProps
                }),
                /* @__PURE__ */ jsx(CalendarWeekHeader, {
                  ...StartCalendarWeekHeaderProps
                }),
                /* @__PURE__ */ jsx(CalendarDateGrid, {
                  ...StartCalendarDataGridProps
                })
              ]
            }),
            /* @__PURE__ */ jsxs(Calendar, {
              selectionVariant: "range",
              ...EndCalendarProps,
              children: [
                /* @__PURE__ */ jsx(CalendarNavigation, {
                  ...EndCalendarNavigationProps
                }),
                /* @__PURE__ */ jsx(CalendarWeekHeader, {
                  ...EndCalendarWeekHeaderProps
                }),
                /* @__PURE__ */ jsx(CalendarDateGrid, {
                  ...EndCalendarDataGridProps
                })
              ]
            })
          ]
        })
      })
    ]
  });
});

export { DatePickerRangePanel };
