// @salt-ds/lab/dist-es/date-picker/DatePickerOverlay.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer, useFloatingComponent, useFormFieldProps, useForkRef } from '@salt-ds/core';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import css_248z from './DatePickerOverlay.css.js';
import { useDatePickerOverlay } from './DatePickerOverlayProvider.js';

const withBaseName = makePrefixer("saltDatePickerOverlay");
const DatePickerOverlay = forwardRef(function DatePickerOverlay2({ className, children, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-date-picker-overlay",
    css: css_248z,
    window: targetWindow
  });
  const { Component: FloatingComponent } = useFloatingComponent();
  const { a11yProps } = useFormFieldProps();
  const {
    state: { open, floatingUIResult },
    helpers: { getFloatingProps }
  } = useDatePickerOverlay();
  const floatingRef = useForkRef(
    ref,
    floatingUIResult == null ? void 0 : floatingUIResult.floating
  );
  return /* @__PURE__ */ jsx(FloatingComponent, {
    className: clsx(withBaseName(), className),
    open: open || false,
    "aria-modal": "true",
    ref: floatingRef,
    focusManagerProps: (floatingUIResult == null ? void 0 : floatingUIResult.context) ? {
      context: floatingUIResult.context,
      initialFocus: 4
    } : void 0,
    ...getFloatingProps ? getFloatingProps({
      ...a11yProps,
      ...rest
    }) : rest,
    children
  });
});

export { DatePickerOverlay };
