// @salt-ds/lab/dist-es/date-picker/DatePicker.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { DateRangeSelectionContext, SingleDateSelectionContext } from './DatePickerContext.js';
import { DatePickerOverlayProvider } from './DatePickerOverlayProvider.js';
import { useDatePicker } from './useDatePicker.js';

const DatePickerMain = forwardRef(
  function DatePickerMain2(props, ref) {
    var _a, _b;
    const {
      children,
      readOnly,
      disabled,
      selectionVariant,
      defaultSelectedDate,
      selectedDate,
      onSelectedDateChange,
      onApply,
      minDate,
      maxDate,
      timeZone,
      locale,
      onCancel,
      ...rest
    } = props;
    const useDatePickerProps = {
      readOnly,
      disabled,
      selectionVariant,
      defaultSelectedDate,
      selectedDate,
      onSelectedDateChange,
      onApply,
      minDate,
      maxDate,
      timeZone,
      locale,
      onCancel
    };
    if (props.selectionVariant === "range") {
      const stateAndHelpers2 = useDatePicker(
        useDatePickerProps,
        ref
      );
      return /* @__PURE__ */ jsx(DateRangeSelectionContext.Provider, {
        value: stateAndHelpers2,
        children: /* @__PURE__ */ jsx("div", {
          ref: (_a = stateAndHelpers2 == null ? void 0 : stateAndHelpers2.state) == null ? void 0 : _a.containerRef,
          ...rest,
          children
        })
      });
    }
    const stateAndHelpers = useDatePicker(
      useDatePickerProps,
      ref
    );
    return /* @__PURE__ */ jsx(SingleDateSelectionContext.Provider, {
      value: stateAndHelpers,
      children: /* @__PURE__ */ jsx("div", {
        ref: (_b = stateAndHelpers == null ? void 0 : stateAndHelpers.state) == null ? void 0 : _b.containerRef,
        ...rest,
        children
      })
    });
  }
);
const DatePicker = forwardRef(
  function DatePicker2(props, ref) {
    const { open, defaultOpen, ...rest } = props;
    return /* @__PURE__ */ jsx(DatePickerOverlayProvider, {
      open,
      defaultOpen,
      children: /* @__PURE__ */ jsx(DatePickerMain, {
        ...rest,
        ref
      })
    });
  }
);

export { DatePicker, DatePickerMain };
