// @salt-ds/lab/dist-cjs/tabs/TabActivationIndicator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var react = require('react');
var useActivationIndicator = require('./useActivationIndicator.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var TabActivationIndicator$1 = require('./TabActivationIndicator.css.js');

const withBaseName = core.makePrefixer("saltTabActivationIndicator");
const TabActivationIndicator = ({
  hideThumb = false,
  orientation = "horizontal",
  tabId
}) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-tab-activation-indicator",
    css: TabActivationIndicator$1,
    window: targetWindow
  });
  const rootRef = react.useRef(null);
  const style = useActivationIndicator.useActivationIndicator({
    rootRef,
    tabId,
    orientation
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    className: withBaseName(),
    ref: rootRef,
    children: hideThumb === false && tabId !== null ? /* @__PURE__ */ jsxRuntime.jsx("div", {
      className: withBaseName("thumb"),
      style
    }) : null
  });
};

exports.TabActivationIndicator = TabActivationIndicator;
