// @salt-ds/lab/dist-cjs/tokenized-input/internal/InputPill.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var react = require('react');
var useWidth = require('./useWidth.js');

const withBaseName = core.makePrefixer("saltInputPill");
const InputPill = react.memo(function InputPill2(props) {
  const {
    active,
    className,
    disabled,
    hidden,
    highlighted,
    index,
    label,
    lastVisible,
    onDelete,
    pillsRef,
    ...restProps
  } = props;
  const ref = react.useRef(null);
  const isRemovable = Boolean(onDelete);
  const { CloseIcon } = core.useIcon();
  core.useIsomorphicLayoutEffect(() => {
    if (!isRemovable && pillsRef.current) {
      pillsRef.current[index] = useWidth.getWidth(ref.current);
    }
  }, [pillsRef, index, isRemovable, lastVisible]);
  core.useIsomorphicLayoutEffect(
    () => () => {
      pillsRef.current[index] = void 0;
    },
    [pillsRef, index]
  );
  const handleDelete = () => {
    onDelete == null ? void 0 : onDelete(index);
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(core.Pill, {
    className: clsx.clsx(
      withBaseName(),
      {
        [withBaseName("pillActive")]: active || highlighted,
        [withBaseName("pillLastVisible")]: lastVisible,
        [withBaseName("hidden")]: hidden
      },
      className
    ),
    disabled,
    tabIndex: -1,
    onClick: isRemovable ? handleDelete : void 0,
    ref,
    role: "option",
    ...restProps,
    children: [
      /* @__PURE__ */ jsxRuntime.jsx("span", {
        className: withBaseName("label"),
        children: label
      }),
      isRemovable && /* @__PURE__ */ jsxRuntime.jsx(CloseIcon, {})
    ]
  });
});

exports.InputPill = InputPill;
