// @salt-ds/lab/dist-es/system-status/SystemStatus.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, StatusIndicator } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './SystemStatus.css.js';

const withBaseName = makePrefixer("saltSystemStatus");
const SystemStatus = forwardRef(
  function SystemStatus2({ children, className, status = "info", ...rest }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-system-status",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxs("div", {
      className: clsx(withBaseName(), withBaseName(status), className),
      ref,
      role: "status",
      ...rest,
      children: [
        /* @__PURE__ */ jsx(StatusIndicator, {
          status,
          className: withBaseName("icon")
        }),
        children
      ]
    });
  }
);

export { SystemStatus };
