// @salt-ds/lab/dist-es/stepper-input/StepperInput.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useIcon, useFormFieldProps, useControlled, useForkRef, capitalize, StatusAdornment, Button } from '@salt-ds/core';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useState, useRef } from 'react';
import { toFixedDecimalPlaces, isOutOfRange, toFloat, isAllowedNonNumeric, sanitizedInput } from './internal/utils.js';
import css_248z from './StepperInput.css.js';
import { useStepperInput } from './useStepperInput.js';

const withBaseName = makePrefixer("saltStepperInput");
const StepperInput = forwardRef(
  function StepperInput2({
    bordered,
    className: classNameProp,
    decimalPlaces = 0,
    defaultValue: defaultValueProp,
    disabled,
    emptyReadOnlyMarker = "\u2014",
    endAdornment,
    hideButtons,
    inputProps: inputPropsProp = {},
    inputRef: inputRefProp,
    max = Number.MAX_SAFE_INTEGER,
    min = Number.MIN_SAFE_INTEGER,
    onChange: onChangeProp,
    placeholder,
    readOnly: readOnlyProp,
    startAdornment,
    step = 1,
    stepBlock = 10,
    textAlign = "left",
    validationStatus: validationStatusProp,
    value: valueProp,
    variant = "primary",
    ...restProps
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-stepper-input",
      css: css_248z,
      window: targetWindow
    });
    const { IncreaseIcon, DecreaseIcon } = useIcon();
    const {
      a11yProps: {
        "aria-describedby": formFieldDescribedBy,
        "aria-labelledby": formFieldLabelledBy
      } = {},
      disabled: formFieldDisabled,
      readOnly: formFieldReadOnly,
      necessity: formFieldRequired,
      validationStatus: formFieldValidationStatus
    } = useFormFieldProps();
    const isDisabled = disabled || formFieldDisabled;
    const isReadOnly = readOnlyProp || formFieldReadOnly;
    const validationStatus = formFieldValidationStatus != null ? formFieldValidationStatus : validationStatusProp;
    const {
      "aria-describedby": inputDescribedBy,
      "aria-labelledby": inputLabelledBy,
      className: inputClassName,
      onBlur: inputOnBlur,
      onChange: inputOnChange,
      onFocus: inputOnFocus,
      required: inputRequired,
      onKeyDown: inputOnKeyDown,
      ...restInputProps
    } = inputPropsProp;
    const isRequired = formFieldRequired ? ["required", "asterisk"].includes(formFieldRequired) : inputRequired;
    const [value, setValue] = useControlled({
      controlled: valueProp,
      default: typeof defaultValueProp === "number" ? toFixedDecimalPlaces(defaultValueProp, decimalPlaces) : defaultValueProp,
      name: "StepperInput",
      state: "value"
    });
    const [focused, setFocused] = useState(false);
    const inputRef = useRef(null);
    const forkedInputRef = useForkRef(inputRef, inputRefProp);
    const {
      decrementButtonProps,
      decrementValue,
      incrementButtonProps,
      incrementValue
    } = useStepperInput({
      inputRef,
      setValue,
      decimalPlaces,
      disabled,
      max,
      min,
      onChange: onChangeProp,
      readOnly: isReadOnly,
      step,
      stepBlock,
      value
    });
    const handleInputFocus = (event) => {
      setFocused(true);
      inputOnFocus == null ? void 0 : inputOnFocus(event);
    };
    const handleInputBlur = (event) => {
      setFocused(false);
      if (value === void 0)
        return;
      const floatValue = toFloat(value);
      if (Number.isNaN(floatValue)) {
        setValue(value);
        onChangeProp == null ? void 0 : onChangeProp(event, value);
      } else {
        const roundedValue = toFixedDecimalPlaces(floatValue, decimalPlaces);
        if (value !== "" && !isAllowedNonNumeric(value)) {
          setValue(roundedValue);
        }
        onChangeProp == null ? void 0 : onChangeProp(event, roundedValue);
      }
      inputOnBlur == null ? void 0 : inputOnBlur(event);
    };
    const handleInputChange = (event) => {
      const changedValue = event.target.value;
      setValue(sanitizedInput(changedValue));
      onChangeProp == null ? void 0 : onChangeProp(event, sanitizedInput(changedValue));
      inputOnChange == null ? void 0 : inputOnChange(event);
    };
    const handleInputKeyDown = (event) => {
      switch (event.key) {
        case "ArrowUp": {
          event.preventDefault();
          const block = event.shiftKey;
          incrementValue(event, block);
          break;
        }
        case "ArrowDown": {
          event.preventDefault();
          const block = event.shiftKey;
          decrementValue(event, block);
          break;
        }
        case "Home": {
          event.preventDefault();
          setValue(min);
          onChangeProp == null ? void 0 : onChangeProp(event, min);
          break;
        }
        case "End": {
          event.preventDefault();
          setValue(max);
          onChangeProp == null ? void 0 : onChangeProp(event, max);
          break;
        }
        case "PageUp": {
          event.preventDefault();
          incrementValue(event, true);
          break;
        }
        case "PageDown": {
          event.preventDefault();
          decrementValue(event, true);
          break;
        }
      }
      inputOnKeyDown == null ? void 0 : inputOnKeyDown(event);
    };
    return /* @__PURE__ */ jsxs("div", {
      className: clsx(withBaseName(), classNameProp),
      ...restProps,
      ref,
      children: [
        /* @__PURE__ */ jsxs("div", {
          className: clsx(
            withBaseName("inputContainer"),
            withBaseName(variant),
            {
              [withBaseName("focused")]: !isDisabled && focused,
              [withBaseName("disabled")]: isDisabled,
              [withBaseName("readOnly")]: isReadOnly,
              [withBaseName(validationStatus || "")]: validationStatus,
              [withBaseName("bordered")]: bordered
            }
          ),
          children: [
            startAdornment && /* @__PURE__ */ jsx("div", {
              className: withBaseName("startAdornmentContainer"),
              children: startAdornment
            }),
            /* @__PURE__ */ jsx("input", {
              "aria-describedby": clsx(formFieldDescribedBy, inputDescribedBy),
              "aria-invalid": isOutOfRange(value, min, max),
              "aria-labelledby": clsx(formFieldLabelledBy, inputLabelledBy),
              "aria-valuemax": toFloat(toFixedDecimalPlaces(max, decimalPlaces)),
              "aria-valuemin": toFloat(toFixedDecimalPlaces(min, decimalPlaces)),
              "aria-valuenow": value && !Number.isNaN(toFloat(value)) ? toFloat(toFixedDecimalPlaces(toFloat(value), decimalPlaces)) : void 0,
              className: clsx(
                withBaseName("input"),
                withBaseName(`inputTextAlign${capitalize(textAlign)}`),
                inputClassName
              ),
              disabled: isDisabled,
              onBlur: handleInputBlur,
              onChange: handleInputChange,
              onFocus: !isDisabled ? handleInputFocus : void 0,
              onKeyDown: handleInputKeyDown,
              placeholder,
              readOnly: isReadOnly,
              ref: forkedInputRef,
              required: isRequired,
              role: "spinbutton",
              tabIndex: isDisabled ? -1 : 0,
              value,
              ...restInputProps
            }),
            !isDisabled && validationStatus && /* @__PURE__ */ jsx(StatusAdornment, {
              status: validationStatus
            }),
            endAdornment && /* @__PURE__ */ jsx("div", {
              className: withBaseName("endAdornmentContainer"),
              children: endAdornment
            }),
            /* @__PURE__ */ jsx("div", {
              className: withBaseName("activationIndicator")
            })
          ]
        }),
        !hideButtons && !isReadOnly && /* @__PURE__ */ jsxs("div", {
          className: withBaseName("buttonContainer"),
          children: [
            /* @__PURE__ */ jsx(Button, {
              className: clsx(
                withBaseName("stepperButton"),
                withBaseName("stepperButtonIncrement")
              ),
              ...incrementButtonProps,
              children: /* @__PURE__ */ jsx(IncreaseIcon, {
                "aria-hidden": true
              })
            }),
            /* @__PURE__ */ jsx(Button, {
              className: clsx(
                withBaseName("stepperButton"),
                withBaseName("stepperButtonDecrement")
              ),
              ...decrementButtonProps,
              children: /* @__PURE__ */ jsx(DecreaseIcon, {
                "aria-hidden": true
              })
            })
          ]
        })
      ]
    });
  }
);

export { StepperInput };
