// @salt-ds/lab/dist-cjs/toolbar/overflow-panel/OverflowPanel.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var ToolbarField = require('../toolbar-field/ToolbarField.js');
var OverflowSeparator = require('./OverflowSeparator.js');
require('../../common-hooks/collectionProvider.js');
require('../../common-hooks/keyUtils.js');
var useCollectionItems = require('../../common-hooks/useCollectionItems.js');
var useKeyboardNavigationPanel = require('../../common-hooks/useKeyboardNavigationPanel.js');
require('../../responsive/useResizeObserver.js');
var DropdownBase = require('../../dropdown/DropdownBase.js');
require('../../dropdown/DropdownButton.js');
require('../../dropdown/Dropdown.js');
require('../../form-field-context-legacy/FormFieldLegacyContext.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var OverflowPanel$1 = require('./OverflowPanel.css.js');

const withBaseName = core.makePrefixer("saltOverflowPanel");
const OverflowPanel = react.forwardRef(function DropdownPanel({
  children,
  className,
  defaultIsOpen,
  isOpen: isOpenProp,
  onOpenChange,
  triggerButtonIcon,
  triggerButtonLabel,
  ...props
}, forwardedRef) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-overflow-panel",
    css: OverflowPanel$1,
    window: targetWindow
  });
  const { OverflowIcon } = core.useIcon();
  const id = core.useIdMemo();
  const collectionHook = useCollectionItems.useCollectionItems({
    id,
    children
  });
  const triggerRef = react.useRef(null);
  const [isOpen, setIsOpen] = core.useControlled({
    controlled: isOpenProp,
    default: defaultIsOpen || false,
    name: "DropdownPanel"
  });
  const { highlightedIndex: highlightedIdx, ...keyboardHook } = useKeyboardNavigationPanel.useKeyboardNavigationPanel({
    cycleFocus: true,
    defaultHighlightedIndex: 0,
    focusOnHighlight: true,
    indexPositions: collectionHook.data
  });
  const focusTrigger = react.useCallback(() => {
    if (triggerRef.current) {
      triggerRef.current.focus();
    }
  }, []);
  const setPanelRef = react.useCallback((el) => el == null ? void 0 : el.focus(), []);
  const handleItemClick = (evt) => {
    setIsOpen(false);
    focusTrigger();
  };
  const handleOpenChange = (open) => {
    setIsOpen(open);
    onOpenChange == null ? void 0 : onOpenChange(open);
  };
  const getTriggerButtonIcon = () => (triggerButtonIcon != null ? triggerButtonIcon : triggerButtonLabel === void 0) ? /* @__PURE__ */ jsxRuntime.jsx(OverflowIcon, {}) : void 0;
  return /* @__PURE__ */ jsxRuntime.jsxs(DropdownBase.DropdownBase, {
    ...props,
    className: clsx.clsx(withBaseName(), className),
    isOpen,
    onOpenChange: handleOpenChange,
    placement: "bottom-end",
    ref: forwardedRef,
    children: [
      /* @__PURE__ */ jsxRuntime.jsxs(core.Button, {
        variant: "secondary",
        ref: triggerRef,
        children: [
          getTriggerButtonIcon(),
          triggerButtonLabel
        ]
      }),
      /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: withBaseName("content"),
        ref: setPanelRef,
        tabIndex: -1,
        ...keyboardHook.listProps,
        children: collectionHook.data.map((item) => {
          const { type } = item.value;
          if (type === OverflowSeparator.OverflowSeparator) {
            return item.value;
          }
          const formFieldProps = {
            id: item.id,
            inOverflowPanel: true,
            key: item.id,
            onClick: handleItemClick
          };
          if (type === ToolbarField.ToolbarField) {
            return react.cloneElement(item.value, formFieldProps);
          }
          return /* @__PURE__ */ react.createElement(ToolbarField.ToolbarField, {
            ...formFieldProps,
            key: item.id
          }, item.value);
        })
      })
    ]
  });
});

exports.OverflowPanel = OverflowPanel;
