// @salt-ds/lab/dist-cjs/calendar/internal/CalendarDay.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var react = require('react');
var useCalendarDay = require('../useCalendarDay.js');
var CalendarDay$1 = require('./CalendarDay.css.js');
var utils = require('./utils.js');

const withBaseName = core.makePrefixer("saltCalendarDay");
const CalendarDay = react.forwardRef(
  function CalendarDay2(props, ref) {
    const {
      className,
      day,
      renderDayContents,
      month,
      TooltipProps,
      formatDate: formatDateProp = utils.formatDate,
      ...rest
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-calendar-day",
      css: CalendarDay$1,
      window: targetWindow
    });
    const dayRef = react.useRef(null);
    const buttonRef = core.useForkRef(ref, dayRef);
    const { status, dayProps, unselectableReason, highlightedReason, locale } = useCalendarDay.useCalendarDay(
      {
        date: day,
        month
      },
      dayRef
    );
    const { outOfRange, today, unselectable, highlighted, hidden, disabled } = status;
    const buttonElement = /* @__PURE__ */ jsxRuntime.jsx("button", {
      "aria-label": formatDateProp(day, locale, {
        day: "2-digit",
        month: "long",
        year: "numeric"
      }),
      disabled,
      type: "button",
      ...dayProps,
      ref: buttonRef,
      ...rest,
      className: clsx.clsx(
        withBaseName(),
        {
          [withBaseName("hidden")]: hidden,
          [withBaseName("outOfRange")]: outOfRange,
          [withBaseName("disabled")]: disabled,
          [withBaseName("unselectable")]: !!unselectable,
          [withBaseName("highlighted")]: !!highlighted
        },
        dayProps.className,
        className
      ),
      children: /* @__PURE__ */ jsxRuntime.jsx("span", {
        className: clsx.clsx(withBaseName("content"), {
          [withBaseName("today")]: today
        }),
        children: renderDayContents ? renderDayContents(day, status) : formatDateProp(day, locale, { day: "numeric" })
      })
    });
    const hasTooltip = unselectableReason || highlightedReason;
    if (!hasTooltip) {
      return buttonElement;
    }
    return /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
      hideIcon: true,
      status: unselectableReason ? "error" : "info",
      content: unselectableReason || highlightedReason || "Date is out of range",
      placement: "top",
      enterDelay: 0,
      leaveDelay: 0,
      ...TooltipProps,
      children: buttonElement
    });
  }
);

exports.CalendarDay = CalendarDay;
