// @salt-ds/core/dist-es/text/Display.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Text } from './Text.js';

const Display1 = forwardRef(function Display12({ children, styleAs = "display1", ...rest }, ref) {
  return /* @__PURE__ */ jsx(Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});
const Display2 = forwardRef(function Display22({ children, styleAs = "display2", ...rest }, ref) {
  return /* @__PURE__ */ jsx(Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});
const Display3 = forwardRef(function Display32({ children, styleAs = "display3", ...rest }, ref) {
  return /* @__PURE__ */ jsx(Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});
const Display4 = forwardRef(function Display42({ children, styleAs = "display4", ...rest }, ref) {
  return /* @__PURE__ */ jsx(Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});

export { Display1, Display2, Display3, Display4 };
