// @salt-ds/lab/dist-cjs/query-input/useQueryInput.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var react = require('react');
require('react/jsx-runtime');
require('../responsive/useResizeObserver.js');
var useWidth = require('../responsive/useWidth.js');
var usePopperStatus = require('./internal/usePopperStatus.js');

function useQueryInput(props, forwardedRef) {
  const bodyRef = react.useRef(null);
  const popperRef = react.useRef(null);
  const inputRef = react.useRef(null);
  const [widthBodyRef, bodyWidth] = useWidth.useWidth(true);
  const forkedRef1 = core.useForkRef(
    forwardedRef,
    bodyRef
  );
  const queryInputBodyRef = core.useForkRef(
    forkedRef1,
    widthBodyRef
  );
  const [selectedItems, setSelectedItems] = core.useControlled({
    controlled: props.selectedItems,
    default: props.defaultSelectedItems || [],
    name: "QueryInput",
    state: "selectedItems"
  });
  const [booleanOperator, setBooleanOperator] = core.useControlled({
    controlled: props.booleanOperator,
    default: props.defaultBooleanOperator || "and",
    name: "QueryInput",
    state: "booleanOperator"
  });
  const popperStatus = usePopperStatus.usePopperStatus({
    initialOpen: false,
    autoClose: props.autoClose
  });
  const [isFocused, setIsFocused] = react.useState(false);
  const [highlightedIndex, setHighlightedIndex] = react.useState(0);
  const [highlightedCategoryIndex, setHighlightedCategoryIndex] = react.useState(0);
  const [highlightedValueIndex, setHighlightedValueIndex] = react.useState(0);
  const [selectedCategory, setSelectedCategory] = react.useState(null);
  const searchListIndexPositions = react.useRef(
    []
  );
  const onInputFocus = (event) => {
    popperStatus.onFocus(event);
    setIsFocused(true);
  };
  const onInputBlur = (event) => {
    var _a, _b;
    if (!inputRef.current) {
      return;
    }
    const relatedTarget = event.relatedTarget;
    const isBodyFocused = bodyRef.current === relatedTarget;
    if (isBodyFocused) {
      inputRef.current.focus();
      return;
    }
    const isFocusWithinBody = (_a = bodyRef.current) == null ? void 0 : _a.contains(relatedTarget);
    if (isFocusWithinBody) {
      return;
    }
    const isFocusWithinPopper = (_b = popperRef.current) == null ? void 0 : _b.contains(relatedTarget);
    if (isFocusWithinPopper) {
      inputRef.current.focus();
      return;
    }
    setIsFocused(false);
    popperStatus.onBlur(event);
  };
  const onFocus = (event) => {
    setIsFocused(true);
    popperStatus.onFocus(event);
  };
  const onBlur = (event) => {
    var _a, _b;
    const relatedTarget = event.relatedTarget;
    const isBodyFocused = bodyRef.current === relatedTarget;
    if (isBodyFocused) {
      return;
    }
    const isFocusWithinBody = (_a = bodyRef.current) == null ? void 0 : _a.contains(relatedTarget);
    if (isFocusWithinBody) {
      return;
    }
    const isFocusWithinPopper = (_b = popperRef.current) == null ? void 0 : _b.contains(relatedTarget);
    if (isFocusWithinPopper) {
      return;
    }
    setIsFocused(false);
    popperStatus.onBlur(event);
  };
  const onSelectedItemsChange = (newItems) => {
    const newItem = newItems == null ? void 0 : newItems.find(
      (item) => item.category == null && item.value === inputValue
    );
    if (newItem) {
      setInputValue("");
    }
    setSelectedItems(newItems || []);
    if (props.onChange) {
      props.onChange(newItems || []);
    }
    popperStatus.onChange();
  };
  const onBooleanOperatorChange = (operator) => {
    setBooleanOperator(operator);
    if (props.onBooleanOperatorChange) {
      props.onBooleanOperatorChange(operator);
    }
  };
  const onSelectedCategoryChange = (category) => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setSelectedCategory(category);
  };
  const [inputValue, setInputValue] = react.useState("");
  const searchListItems = react.useMemo(() => {
    const [searchListItems2, indexPositions] = filterCategories(
      props.categories,
      inputValue
    );
    searchListIndexPositions.current = indexPositions;
    return searchListItems2;
  }, [props.categories, inputValue]);
  const onInputChange = (event) => {
    const newInputValue = event.target.value;
    setInputValue(newInputValue);
    setHighlightedIndex(searchListIndexPositions.current.length === 0 ? 0 : 1);
  };
  const searchListItemCount = react.useMemo(() => {
    return searchListItems.reduce(
      (acc, category) => acc + category.values.length,
      0
    );
  }, [searchListItems]);
  const onInputClear = () => {
    onSelectedItemsChange([]);
  };
  const onInputClick = () => {
    popperStatus.onClick();
  };
  const newItemFromSelected = () => {
    const i = highlightedIndex;
    for (const category of searchListItems) {
      if (i < category.values.length) {
        const value = category.values[i];
        return {
          category: category.name,
          value
        };
      }
    }
  };
  const { displayedItemCount = 10 } = props;
  const onSearchListKeyDown = (key) => {
    switch (key) {
      case "Home":
        setHighlightedIndex(0);
        return;
      case "End":
        setHighlightedIndex(searchListItemCount);
        return;
      case "ArrowUp":
        setHighlightedIndex(
          (i) => prevSearchItemIndex(i, searchListIndexPositions.current)
        );
        return;
      case "ArrowDown":
        setHighlightedIndex(
          (i) => nextSearchItemIndex(i, searchListIndexPositions.current)
        );
        return;
      case "PageDown":
        setHighlightedIndex(
          (i) => Math.min(searchListItemCount, i + displayedItemCount)
        );
        return;
      case "PageUp":
        setHighlightedIndex((i) => Math.max(0, i - displayedItemCount));
        return;
      case "Enter":
        if (highlightedIndex === searchListItemCount) {
          const newItems = [
            ...selectedItems,
            { category: null, value: inputValue }
          ];
          onSelectedItemsChange(newItems);
        } else {
          if (highlightedIndex >= 0 && highlightedIndex < searchListItemCount) {
            const newItem = newItemFromSelected();
            if (newItem) {
              const newItems = [...selectedItems, newItem];
              onSelectedItemsChange(newItems);
            }
          }
          setInputValue("");
        }
        return;
      default:
        return;
    }
  };
  const onCategoryListKeyDown = (key) => {
    switch (key) {
      case "Home":
        setHighlightedCategoryIndex(0);
        return;
      case "End":
        setHighlightedCategoryIndex(props.categories.length - 1);
        return;
      case "ArrowUp":
        setHighlightedCategoryIndex((i) => Math.max(0, i - 1));
        return;
      case "ArrowDown":
        setHighlightedCategoryIndex(
          (i) => Math.min(props.categories.length - 1, i + 1)
        );
        return;
      case "PageUp":
        setHighlightedCategoryIndex((i) => Math.max(0, i - displayedItemCount));
        return;
      case "PageDown":
        setHighlightedCategoryIndex(
          (i) => Math.min(props.categories.length - 1, i + displayedItemCount)
        );
        return;
      case "Enter":
        if (highlightedCategoryIndex >= 0 && highlightedCategoryIndex < props.categories.length) {
          const category = props.categories[highlightedCategoryIndex];
          setSelectedCategory(category);
        }
        return;
      case "ArrowRight":
        if (highlightedCategoryIndex >= 0 && highlightedCategoryIndex < props.categories.length) {
          const category = props.categories[highlightedCategoryIndex];
          setSelectedCategory(category);
        }
        return;
      default:
        return;
    }
  };
  const onValueListKeyDown = (key) => {
    switch (key) {
      case "Home":
        setHighlightedValueIndex(0);
        return;
      case "End":
        if (selectedCategory != null) {
          setHighlightedValueIndex(selectedCategory.values.length);
        }
        return;
      case "ArrowUp":
        setHighlightedValueIndex((i) => Math.max(0, i - 1));
        return;
      case "ArrowDown":
        setHighlightedValueIndex((i) => {
          if (selectedCategory != null) {
            return Math.min(selectedCategory.values.length, i + 1);
          }
          return i;
        });
        return;
      case "PageUp":
        setHighlightedValueIndex((i) => Math.max(0, i - displayedItemCount));
        return;
      case "PageDown":
        setHighlightedValueIndex((i) => {
          if (selectedCategory != null) {
            return Math.min(
              selectedCategory.values.length,
              i + displayedItemCount
            );
          }
          return i;
        });
        return;
      case "Enter":
        if (highlightedValueIndex === 0 || selectedCategory == null) {
          setSelectedCategory(null);
        } else {
          const value = selectedCategory.values[highlightedValueIndex - 1];
          const newItems = selectedItems.filter(
            (item) => !(item.category === selectedCategory.name && item.value === value)
          );
          if (newItems.length === selectedItems.length) {
            newItems.push({ category: selectedCategory.name, value });
          }
          onSelectedItemsChange(newItems);
        }
        return;
      case "ArrowLeft":
        setSelectedCategory(null);
        return;
      default:
        return;
    }
  };
  const onKeyDown = (event) => {
    popperStatus.onKeyDown(event);
    let key = event.key;
    if (event.key === " " && event.ctrlKey) {
      key = "Enter";
    }
    const isSearchListActive = inputValue !== "";
    const isCategoryListActive = !isSearchListActive && !selectedCategory;
    const isValueListActive = !isSearchListActive && !isCategoryListActive;
    if (isSearchListActive) {
      onSearchListKeyDown(key);
    } else if (isCategoryListActive) {
      onCategoryListKeyDown(key);
    } else {
      onValueListKeyDown(key);
    }
    switch (event.key) {
      case "ArrowUp":
      case "ArrowDown":
      case "Enter":
        event.preventDefault();
        break;
      case "ArrowLeft":
        if (isValueListActive) {
          event.preventDefault();
        }
        break;
      case "ArrowRight":
        if (isCategoryListActive) {
          event.preventDefault();
        }
        break;
      case "Escape":
        setInputValue("");
    }
  };
  const onValueToggle = (category, value) => {
    const newItems = selectedItems.filter(
      (item) => !(item.category === category.name && item.value === value)
    );
    if (newItems.length === selectedItems.length) {
      newItems.push({ category: category.name, value });
    }
    onSelectedItemsChange(newItems);
  };
  return {
    queryInputBodyRef,
    queryInputProps: {
      onFocus,
      onBlur
    },
    queryInputBodyProps: {
      inputRef,
      disabled: props.disabled,
      selectedItems,
      onFocus,
      onBlur,
      onInputClick,
      onInputFocus,
      onInputBlur,
      onInputClear,
      inputValue,
      onInputChange,
      onSelectedItemsChange,
      onKeyDown,
      isFocused,
      booleanOperator,
      onBooleanOperatorChange
    },
    valueSelectorProps: {
      isOpen: popperStatus.isOpen,
      categories: props.categories,
      selectedCategory,
      selectedItems,
      onValueToggle,
      onSearchListChange: onSelectedItemsChange,
      anchorElement: bodyRef.current,
      width: bodyWidth,
      onSelectedCategoryChange,
      popperRef,
      inputValue,
      highlightedIndex,
      visibleCategories: searchListItems,
      highlightedCategoryIndex,
      highlightedValueIndex,
      setHighlightedCategoryIndex,
      setHighlightedValueIndex,
      setHighlightedIndex
    }
  };
}
function filterCategories(categories, inputValue) {
  if (!inputValue) {
    return [categories, []];
  }
  const query = inputValue.toUpperCase();
  const visibleCategories = [];
  const indexPositions = [];
  for (const c of categories) {
    const values = c.values.filter((v) => v.toUpperCase().includes(query));
    if (values.length > 0) {
      const queryInputCategory = {
        name: c.name,
        values
      };
      visibleCategories.push(queryInputCategory);
      indexPositions.push(queryInputCategory, ...values);
    }
  }
  return [visibleCategories, indexPositions];
}
function nextSearchItemIndex(index, indexPositions) {
  const nextIndex = index + 1;
  if (nextIndex === indexPositions.length + 1) {
    return index;
  }
  if (nextIndex === indexPositions.length) {
    return nextIndex;
  }
  if (typeof indexPositions[nextIndex] === "string") {
    return nextIndex;
  }
  return nextIndex + 1;
}
function prevSearchItemIndex(index, indexPositions) {
  const nextIndex = index - 1;
  if (nextIndex === 0) {
    return index;
  }
  if (typeof indexPositions[nextIndex] === "string") {
    return nextIndex;
  }
  return nextIndex - 1;
}

exports.useQueryInput = useQueryInput;
