// @salt-ds/lab/dist-es/calendar/CalendarWeekHeader.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { daysForLocale } from './internal/utils.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './CalendarWeekHeader.css.js';
import { getCurrentLocale } from './formatDate.js';

const withBaseName = makePrefixer("saltCalendarWeekHeader");
const CalendarWeekHeader = forwardRef(function CalendarWeekHeader2({ className, locale = getCurrentLocale(), ...rest }, ref) {
  const weekdaysShort = daysForLocale("narrow", locale);
  const weekdaysLong = daysForLocale("long", locale);
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-calendar-week-header",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx("div", {
    "data-testid": "CalendarWeekHeader",
    className: clsx(withBaseName(), className),
    ref,
    ...rest,
    children: weekdaysShort.map((day, index) => /* @__PURE__ */ jsx("small", {
      "aria-hidden": true,
      className: withBaseName("dayOfWeek"),
      children: day
    }, weekdaysLong[index]))
  });
});

export { CalendarWeekHeader };
