// @salt-ds/lab/dist-cjs/list-deprecated/internal/Highlighter.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
require('react');
var escapeRegExp = require('../../utils/escapeRegExp.js');
require('@salt-ds/core');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Highlighter$1 = require('./Highlighter.css.js');

const baseName = "saltHighlighter";
const Highlighter = (props) => {
  const { matchPattern, text = "" } = props;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-highligher-deprecated",
    css: Highlighter$1,
    window: targetWindow
  });
  const matchRegex = typeof matchPattern === "string" ? new RegExp(`(${escapeRegExp.escapeRegExp(matchPattern)})`, "gi") : matchPattern;
  return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: text.split(matchRegex).map(
      (part, index) => part.match(matchRegex) ? /* @__PURE__ */ jsxRuntime.jsx("strong", {
        className: `${baseName}-highlight`,
        children: part
      }, `${index}-${part}`) : part
    )
  });
};

exports.Highlighter = Highlighter;
