// @salt-ds/lab/dist-cjs/color-chooser/RGBAInputField.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var react = require('react');
var InputLegacy = require('../input-legacy/InputLegacy.js');
require('../input-legacy/StaticInputAdornment.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var RGBAInput = require('./RGBAInput.css.js');

const withBaseName = core.makePrefixer("saltColorChooser");
const RGBInput = ({
  rgbaValue,
  value,
  onSubmit
}) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-rgba-input",
    css: RGBAInput,
    window: targetWindow
  });
  const [rgbaInputValue, setRgbaInputValue] = react.useState(
    rgbaValue ? rgbaValue[value] : ""
  );
  react.useEffect(() => {
    setRgbaInputValue(rgbaValue ? rgbaValue[value] : "");
  }, [rgbaValue, value]);
  const handleRGBInputChange = (e, value2) => {
    let rgb;
    rgb = Number.parseInt(value2);
    if (value2.trim() === "" || Number.isNaN(rgb)) {
      rgb = "";
    }
    setRgbaInputValue(rgb);
  };
  const handleKeyDownRGB = (e) => {
    if (e.key === "Enter") {
      const newRgb = { ...rgbaValue, [value]: e.currentTarget.value };
      const validatedRgb = {
        r: Math.max(0, Math.min(newRgb.r, 255)),
        g: Math.max(0, Math.min(newRgb.g, 255)),
        b: Math.max(0, Math.min(newRgb.b, 255)),
        a: newRgb.a
      };
      onSubmit(validatedRgb);
    }
  };
  const handleOnBlurRGB = (e) => {
    const newRgb = { ...rgbaValue, [value]: e.target.value };
    const validatedRgb = {
      r: Math.max(0, Math.min(newRgb.r, 255)),
      g: Math.max(0, Math.min(newRgb.g, 255)),
      b: Math.max(0, Math.min(newRgb.b, 255)),
      a: newRgb.a
    };
    onSubmit(validatedRgb, e);
  };
  return /* @__PURE__ */ jsxRuntime.jsx(InputLegacy.InputLegacy, {
    inputProps: {
      "data-testid": `${value}-input`
    },
    className: withBaseName("rgbaInput"),
    value: rgbaInputValue.toString(),
    onChange: handleRGBInputChange,
    onBlur: handleOnBlurRGB,
    onKeyDown: handleKeyDownRGB
  });
};

exports.RGBInput = RGBInput;
