// @salt-ds/lab/dist-cjs/system-status/SystemStatus.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var react = require('react');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var SystemStatus$1 = require('./SystemStatus.css.js');

const withBaseName = core.makePrefixer("saltSystemStatus");
const SystemStatus = react.forwardRef(
  function SystemStatus2({ children, className, status = "info", ...rest }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-system-status",
      css: SystemStatus$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsxs("div", {
      className: clsx.clsx(withBaseName(), withBaseName(status), className),
      ref,
      role: "status",
      ...rest,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(core.StatusIndicator, {
          status,
          className: withBaseName("icon")
        }),
        children
      ]
    });
  }
);

exports.SystemStatus = SystemStatus;
