// @salt-ds/lab/dist-cjs/dropdown/DropdownBase.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var react$1 = require('@floating-ui/react');
var core = require('@salt-ds/core');
var styles = require('@salt-ds/styles');
var clsx = require('clsx');
var react = require('react');
var Portal = require('../portal/Portal.js');
var forwardCallbackProps = require('../utils/forwardCallbackProps.js');
var WindowContext = require('../window/WindowContext.js');
require('../window/ElectronWindow.js');
var useDropdownBase = require('./useDropdownBase.js');
var window = require('@salt-ds/window');
var Dropdown = require('./Dropdown.css.js');

const withBaseName = core.makePrefixer("saltDropdownBase");
const DropdownBase = react.forwardRef(
  function Dropdown$1({
    "aria-labelledby": ariaLabelledByProp,
    children,
    className: classNameProp,
    container,
    defaultIsOpen,
    disabled,
    disablePortal,
    fullWidth,
    id: idProp,
    isOpen: isOpenProp,
    onKeyDown,
    onOpenChange,
    openOnFocus,
    placement = "bottom-start",
    popupWidth,
    width,
    ...htmlAttributes
  }, forwardedRef) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-dropdown",
      css: Dropdown,
      window: targetWindow
    });
    const rootRef = react.useRef(null);
    const className = clsx.clsx(withBaseName(), classNameProp, {
      [withBaseName("fullWidth")]: fullWidth,
      [withBaseName("disabled")]: disabled
    });
    const [trigger, popupComponent] = react.Children.toArray(
      children
    );
    const id = core.useIdMemo(idProp);
    const Window = WindowContext.useWindow();
    const { componentProps, popperRef, isOpen, triggerProps } = useDropdownBase.useDropdownBase(
      {
        ariaLabelledBy: ariaLabelledByProp,
        defaultIsOpen,
        disabled,
        fullWidth,
        id,
        isOpen: isOpenProp,
        onOpenChange,
        onKeyDown,
        openOnFocus,
        popupComponent,
        popupWidth,
        rootRef,
        width
      }
    );
    const [maxPopupHeight, setMaxPopupHeight] = react.useState(
      void 0
    );
    const middleware = WindowContext.isDesktop ? [] : [
      react$1.flip({
        fallbackPlacements: ["bottom-start", "top-start"]
      }),
      react$1.shift({ limiter: react$1.limitShift() }),
      react$1.size({
        apply({ availableHeight }) {
          setMaxPopupHeight(availableHeight);
        }
      })
    ];
    const { reference, floating, x, y, strategy } = core.useFloatingUI({
      placement,
      middleware
    });
    const handlePopperListAdapterRef = core.useForkRef(
      reference,
      forwardedRef
    );
    const handleRootRef = core.useForkRef(rootRef, handlePopperListAdapterRef);
    const handleFloatingRef = core.useForkRef(floating, popperRef);
    const getTriggerComponent = () => {
      const {
        id: defaultId,
        role: defaultRole,
        ...restTriggerProps
      } = triggerProps;
      const {
        id: id2 = defaultId,
        role = defaultRole,
        ...ownProps
      } = trigger.props;
      return react.cloneElement(
        trigger,
        forwardCallbackProps.forwardCallbackProps(ownProps, {
          ...restTriggerProps,
          id: id2,
          role
        })
      );
    };
    const getPopupComponent = () => {
      const { id: defaultId, width: width2, ...restComponentProps } = componentProps;
      const {
        className: className2,
        id: id2 = defaultId,
        width: ownWidth,
        ...ownProps
      } = popupComponent.props;
      return react.cloneElement(popupComponent, {
        ...ownProps,
        ...restComponentProps,
        className: clsx.clsx(className2, withBaseName("popup-component")),
        id: id2,
        width: ownWidth != null ? ownWidth : width2
      });
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", {
      ...htmlAttributes,
      className,
      "data-testid": "dropdown",
      id: idProp,
      ref: handleRootRef,
      children: [
        getTriggerComponent(),
        isOpen && /* @__PURE__ */ jsxRuntime.jsx(Portal.Portal, {
          disablePortal,
          container,
          children: /* @__PURE__ */ jsxRuntime.jsx(Window, {
            className: clsx.clsx(withBaseName("popup"), classNameProp),
            id: `${id}-popup`,
            style: {
              top: y != null ? y : 0,
              left: x != null ? x : 0,
              position: strategy,
              maxHeight: maxPopupHeight
            },
            ref: handleFloatingRef,
            children: getPopupComponent()
          })
        })
      ]
    });
  }
);

exports.DropdownBase = DropdownBase;
