// @salt-ds/lab/dist-cjs/dropdown/useDropdownBase.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var react = require('react');
require('../form-field-context-legacy/FormFieldLegacyContext.js');
var useFormFieldLegacyProps = require('../form-field-context-legacy/useFormFieldLegacyProps.js');
require('react/jsx-runtime');
var useResizeObserver = require('../responsive/useResizeObserver.js');
var useClickAway = require('./useClickAway.js');

const NO_OBSERVER = [];
const useDropdownBase = ({
  ariaLabelledBy: ariaLabelledByProp,
  defaultIsOpen,
  disabled,
  fullWidth: fullWidthProp,
  id,
  isOpen: isOpenProp,
  onOpenChange,
  onKeyDown: onKeyDownProp,
  openOnFocus,
  popupWidth: popupWidthProp,
  rootRef,
  width
}) => {
  var _a;
  const justFocused = react.useRef(null);
  const popperRef = react.useRef(null);
  const popperCallbackRef = react.useCallback((element) => {
    popperRef.current = element;
  }, []);
  const [isOpen, setIsOpen] = core.useControlled({
    controlled: isOpenProp,
    default: Boolean(defaultIsOpen),
    name: "useDropdown",
    state: "isOpen"
  });
  const {
    inFormField,
    setFocused: setFormFieldFocused,
    a11yProps: { "aria-labelledby": ariaLabelledBy, ...restA11yProps } = {}
  } = useFormFieldLegacyProps.useFormFieldLegacyProps();
  const [popup, setPopup] = react.useState({
    width: (_a = popupWidthProp != null ? popupWidthProp : width) != null ? _a : 0
  });
  const showDropdown = react.useCallback(() => {
    setIsOpen(true);
    onOpenChange == null ? void 0 : onOpenChange(true);
  }, [onOpenChange]);
  const hideDropdown = react.useCallback(() => {
    setIsOpen(false);
    onOpenChange == null ? void 0 : onOpenChange(false);
  }, [onOpenChange]);
  useClickAway.useClickAway({
    popperRef,
    rootRef,
    isOpen,
    onClose: hideDropdown
  });
  const handleTriggerFocus = react.useCallback(() => {
    if (!disabled) {
      setFormFieldFocused == null ? void 0 : setFormFieldFocused(true);
      if (openOnFocus) {
        setIsOpen(true);
        onOpenChange == null ? void 0 : onOpenChange(true);
        justFocused.current = window.setTimeout(() => {
          justFocused.current = null;
        }, 1e3);
      }
    }
  }, [disabled, onOpenChange, openOnFocus, setFormFieldFocused]);
  const handleTriggerBlur = react.useCallback(() => {
    setFormFieldFocused == null ? void 0 : setFormFieldFocused(false);
  }, [setFormFieldFocused]);
  const handleTriggerToggle = react.useCallback(
    (e) => {
      if (["Enter", " "].indexOf(
        e.key
      ) === -1) {
        const newIsOpen = !isOpen;
        setIsOpen(newIsOpen);
        onOpenChange == null ? void 0 : onOpenChange(newIsOpen);
      }
    },
    [isOpen, onOpenChange]
  );
  const handleKeydown = react.useCallback(
    (evt) => {
      if ((evt.key === "Tab" || evt.key === "Escape") && isOpen) {
        hideDropdown();
      } else if ((evt.key === "Enter" || evt.key === "ArrowDown" || evt.key === " ") && !isOpen) {
        evt.preventDefault();
        showDropdown();
      } else {
        onKeyDownProp == null ? void 0 : onKeyDownProp(evt);
      }
    },
    [hideDropdown, isOpen, onKeyDownProp, showDropdown]
  );
  const fullWidth = fullWidthProp != null ? fullWidthProp : inFormField;
  const measurements = fullWidth ? useResizeObserver.WidthOnly : NO_OBSERVER;
  useResizeObserver.useResizeObserver(rootRef, measurements, setPopup, fullWidth);
  const componentId = `${id}-dropdown`;
  const getAriaLabelledBy = (labelledBy, labelledByProp) => {
    if (labelledBy === void 0 && labelledByProp === void 0) {
      return void 0;
    }
    return [labelledBy, labelledByProp].filter((x) => !!x).join(" ");
  };
  const triggerProps = {
    ...restA11yProps,
    "aria-expanded": isOpen,
    "aria-labelledby": getAriaLabelledBy(ariaLabelledBy, ariaLabelledByProp),
    "aria-owns": isOpen ? componentId : void 0,
    id: `${id}-control`,
    onClick: disabled || openOnFocus ? void 0 : handleTriggerToggle,
    onFocus: handleTriggerFocus,
    onBlur: handleTriggerBlur,
    role: "listbox",
    onKeyDown: disabled ? void 0 : handleKeydown,
    style: { width: fullWidth ? void 0 : width }
  };
  const dropdownComponentProps = {
    "aria-labelledby": ariaLabelledBy,
    id: componentId,
    width: popup.width
  };
  return {
    componentProps: dropdownComponentProps,
    popperRef: popperCallbackRef,
    isOpen,
    label: "Dropdown Button",
    triggerProps
  };
};

exports.useDropdownBase = useDropdownBase;
