// @salt-ds/lab/dist-cjs/menu-button/MenuButtonTrigger.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var react = require('react');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var MenuButtonTrigger$1 = require('./MenuButtonTrigger.css.js');

const withBaseName = core.makePrefixer("saltMenuButtonTrigger");
const MenuButtonTrigger = react.forwardRef(function MenuButtonTrigger2({ className, hideCaret, isMenuOpen, children, ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-menu-button-trigger",
    css: MenuButtonTrigger$1,
    window: targetWindow
  });
  const { ExpandIcon } = core.useIcon();
  return /* @__PURE__ */ jsxRuntime.jsxs(core.Button, {
    ref,
    className: clsx.clsx(withBaseName(), className, {
      [withBaseName("buttonOpen")]: isMenuOpen
    }),
    "data-testid": "menu-button-trigger",
    ...rest,
    children: [
      children,
      !hideCaret && /* @__PURE__ */ jsxRuntime.jsx(ExpandIcon, {
        className: withBaseName("caretIcon"),
        "data-testid": "menu-button-trigger-caret"
      })
    ]
  });
});

exports.MenuButtonTrigger = MenuButtonTrigger;
